### ISSUE TITLE
#[repr(C)] silently accepted on ADT enums

### ISSUE THREAD

**Issue by durka** [rust-lang/rust#31692]
This code compiles without error or relevant warning:

``` rust
#[repr(C)]
enum Foo { A { x: i32 }, B(i32), C }

#[repr(i32)]
enum Bar { Bar(i32), Baz(u32) }
```

But obviously the representation of the enum is still going to be undefined, and not integral or C-compatible (trivially checked by `mem::size_of` in this case).

There should be an error (probably infeasible) or at least a lint that is on by default.

@aatch says there is already an issue about this, but I can't find it, so feel free to close as a duplicate if you can find it :)
cc @niconii

**Comment by nodakai** [comment]
Take a look at #25651

**Comment by durka** [comment]
Okay, so `#[repr]` on an enum controls the size of the discriminant, not the whole enum. At least `#[repr(C)]` should warn IMO because `#[repr(C)]` means "give this thing a defined layout", but that doesn't happen in this case.
